### PR TITLE
added part from SD fat lib which enables SPI 2 device write access

### PR DESCRIPTION
**Pull request by stevstrong**
taken from SdSpiSTM32F1.cpp - Victor's lib, and adapted to support device selection.
To be checked with other libs which are using the SPI lib.

### CODE REVIEW

**Comment by stevstrong** [comment]
removes unwanted glitches before getting SCK active caused by setting parameters after begin().

**Comment by rogerclarkmelbourne** [comment]
Hi Steve

Sorry, I've not actioned this, but I've had quite a lot of other PR's and work on the go.

With your change I need to compile and test it on several devices, so it will take time before I can confirm its OK and action the PR.

**Comment by stevstrong** [comment]
No problem, Roger, I understand.
I just wonder that nobody tried to use SPI 2 before so that this issue wasn't reported yet.
Anyway, I am using it successfully since then, so side effects so far. :)

**Comment by rogerclarkmelbourne** [comment]
Steve

I recall doing some testing using SPI2 with a mp3 audio playback module on 2 and SD on 1,and I know it was working before I added the setModule stuff.

However the SD lib at that time was rather flakey and I had issues of it occasionally timing out during reads for no apparent reason.
This could have been caused by the problems you have found with spikes on the lines, but as the SD lib had been recently modified by @victorpv and the author, I thought there were still bugs in it.

And as the mp3 playback was just done for a matter of interest, I parked the project on a shelf ;-)

PS. feel free to post to the PR, I think this is normal

**Comment by stevstrong** [comment]
OK, so if only mp3 playback was tested, no __writing__ took place, the writing problem could probably be still there.

**Comment by rogerclarkmelbourne** [comment]
I think it was reading from SPI1 and writing to SPI2, but I can't be absolutely sure.

I recall other people saying they were going to try the setModule() stuff, but I can't recall if anyone did

As this is a community driven project, this often only get tested for what people use them for.
Consequentially, I often get PRs that work fine for one person, but end up breaking stuff for other people

Or I get PR's which seem Ok because either I or the person who developed the change know it works for us, but don't have the time to test for all possible use of the change etc

So we are always going to end up with some bugs, which need to be addressed when someone tried to use stuff that's not been tested in that usage before.

Its just the nature of the project

**Comment by rogerclarkmelbourne** [comment]
Steve,

The SPI.cpp does not compile

You seem to have dashes at the start of lines 352 etc  e.g.

- // taken from SdSpiSTM32F1.cpp - Victor's lib, and adapted to support device selection
- if (spi_is_rx_nonempty(_currentSetting->spi_d)) {
- uint8_t b = spi_rx_reg(_currentSetting->spi_d);
- }

Have you committed you latest version ?

I could try removing these, but I'm a bit concerned that you may have other uncommited / pushed changes

**Comment by rogerclarkmelbourne** [comment]
Steve.

You also have an issue with your ADC changes

You have removed

#define ADC_CR2_EXTSEL 0xE0000

but this is used by adc.cpp in the core

Perhaps you have not pushed the adc stuff either ?

**Comment by stevstrong** [comment]
Regarding __SPI.cpp__ does not compile

    // taken from SdSpiSTM32F1.cpp - Victor's lib, and adapted to support device selection
    if (spi_is_rx_nonempty(_currentSetting->spi_d)) {
        uint8_t b = spi_rx_reg(_currentSetting->spi_d);
    }
I have to double-check it, I'll be back soon.

**Comment by stevstrong** [comment]
Well, somehow are the dashes there, don´t know how...
I removed them locally and it compiles now without errors.
Could you please delete those dashes, I somehow cannot propose any more changes to the sources...

**Comment by rogerclarkmelbourne** [comment]
Hi Steve

I'll try pulling again from your repo and then locally delete those dashes

BTW, you can edit a file on github (but them you need to pull locally otherwise you can't push)

**Comment by stevstrong** [comment]
maybe it will work now.

**Comment by rogerclarkmelbourne** [comment]
Thanks steve

I'll try it again

**Comment by rogerclarkmelbourne** [comment]
Steve

Pulling and merging your master branch has brought all of your changes into my development branch, which I've now pushed back to github

When you get chance, can you re-test against my development branch

Thanks

Roger

**Comment by rogerclarkmelbourne** [comment]
Steve

I've forgotten where we were with this

I think I did the changes manually, or are there other changes that still need to be pulled ??

**Comment by stevstrong** [comment]
Hi Roger, I was lately busy with other tasks, but I will check the status on this week-end.

**Comment by rogerclarkmelbourne** [comment]
Steve

There is no rush.

I just noticed this issue was still open, and couldnt remember if it had been fixed by an alternative means; as I did a load of manual updates to the files, based on your changes - so its likely that everything is OK and I can close this.

**Comment by phenom128** [comment]
roger,
is it possible post a 'summary' subject in the 'libraries working' forum
for the SD library?
it's rather confusing and spread out across forum posts as to just what
bits are where and their precise usability status.

it would probably you help to decide if the issues are now resolved:-)
if you want me to instead raise a forum topic i can.

stephen

**Comment by rogerclarkmelbourne** [comment]
Steve

I don't keep track of libraries, I simply wouldn't have the time to do this.

Some people send me PR's with new libraries in them, to include into the repo, but most people don't seem to have a github account, so their either post zips to the forum or links to zips on their own website.

Even when I get some free time, Its only possible for me to test a small subset of the libraries, as I only have a small subset of the hardware that other people use.

**Comment by phenom128** [comment]
ok, fair enough.

i'll see if i can figure it out.

stephen

**Comment by stevstrong** [comment]
Hi Roger,
I finally managed to check this issue. But it wasn't that smooth as I though it would be.
I loaded locally your master branch merged with my one, compiled, loaded... the SD card access (on SPI2, using ``SPI.setModule(2);``) failed. Again, it was the scope probe effect, only worked if I touched the SCK line with the probe. It's true, the HW is different from that I used before, but it was still annoying.
So I had to rework again the SPI access functions, and managed to get this time an SCK line entirely without any glitches! As a result, my SD card is working even with 18MHz clock! And, of course, it works now without any probe and without any additional filtering capacitors.
I can say, this time it looks really promising.
Hopefully other users can acknowledge this.
And here is the PR.

**Comment by rogerclarkmelbourne** [comment]
Steve is this a pull to the development or the master branch, as I have a feeling most of your stuff went into the development branch.

Actually, I'm tempted to merge the development branch back into the master as hardly anyone uses the development branch and I don't want it to get to far ahead of the master.

But the problem, is that as hardly anyone updates on a regular basis, (either the master or the development branch), nothing gets a large amount of testing :-(

**Comment by stevstrong** [comment]
Hi Roger, I understand, I agree with you to merge development to master branch, the people who use this SW will test it this way :). And there is no big deal even if something will not work (what I don't expect), the changes are quite transparent, and I am currently active and would fix my part a.s.a.p.

**Comment by stevstrong** [comment]
OK, I figured out. I used your development branch from one month ago, then merged with the master branch 15 days ago (with 49 changes), and another merge with master 15 days ago (7 changes).
So it looks like I used the master+development combination. Which is good, isn't it?

**Comment by rogerclarkmelbourne** [comment]
Hi Steve,

That sounds about right.

Was SD still working after you did all that merging ?

**Comment by stevstrong** [comment]
Stop. I have a problem.
I just realize I have a ... hm ... compiler (?) problem.
Together with SD card writing/reading, I use this:
``typedef struct {
uint8_t byte0;
uint16_t len;
uint16_t crc;
} pack_t;``
If I printout ``sizeof(pack_t)``, I would expect 5, but I get value 6. What am I doing wrong?
Sorry, maybe this is not the right place to post this issue, but is influencing my testing on this point, because I observe strange things, sometimes works the SD writing, sometimes not...

**Comment by avikde** [comment]
One suggestion is the packed attribute

```
typedef struct {
  //
} __attribute__ ((packed)) Blah;
```

**Comment by stevstrong** [comment]
Thanks, I will try this.
But my concern is that the SdFat lib and other libs may don't have this attribute set, so when I compile them they output can give unpredictable results...
Is there any general solution for that? A compiler switch or so?

**Comment by sweetlilmre** [comment]
My 2c:

A global compiler switch is a bad idea. Structure packing is something that
should be explicitly stated for data on the wire or that represents data
from physical media (your case I'm assuming).

If this was not explicitly stated in the library and you inherited it (this
was not your code) then that was a potential oversight, otherwise you
should specify it. Packing can introduce unaligned performance hits or even
faults (old ARM processors faulted on unaligned access).

In summary: if you need to represent a "physical" byte stream as a
structure, explicitly specify packing so that there can be no confusion.

-(e)

**Comment by rogerclarkmelbourne** [comment]
I agree with @sweetlilmre 

Changing compiler options has proved very problematic in the past as it may fix one issue, but often causes 10 other issues in the process.

**Comment by stevstrong** [comment]
Don't understand me wrong, I don't insist to change something (especially a thing which can globaly influence the code), I'm just wondering why is this issue coming up now and why didn't I observed before? Maybe because I was working mainly on 8 bit processors? There, I had never such kind of issues, and never used this attribute, I didn't even know that it exists.

Anyway, me, I could use packing, no problem, but what about the SDFat lib? I didn't see any use of this attribute there, but I saw a lot of structures... Shall I go through and check each struct? And make it packed?

**Comment by rogerclarkmelbourne** [comment]
If there are issues with the actual SD lib, the author / maintainer is very helpful and has made changes for us in the past. 

I forget exactly who maintains the master repo for the SD lib, but I know @victorpv knows.

I think its best to PM him about this, so he can give you the contact details

**Comment by rogerclarkmelbourne** [comment]
I think this PR has been superseded by my merge of development into Master

If it still applies, can you resend the PR after first pulling this Master
